pkg/trie/triedb/proof/generate.go

```diff
@@ -0,0 +1,452 @@
+	}
+)
+
+func (nodeHandleHash) isNodeHandle()   {}
+func (nodeHandleInline) isNodeHandle() {}
+
+type step interface {
+	isProofStep()
+}
+
+type (
+	stepDescend struct {
+		childPrefixLen int
+		child          nodeHandle
+	}
+	stepFoundValue struct {
+		value *[]byte
+	}
+	stepFoundHashedValue struct {
+		hash []byte
+	}
+)
+
+func (stepDescend) isProofStep()          {}
+func (stepFoundValue) isProofStep()       {}
+func (stepFoundHashedValue) isProofStep() {}
+
+type stackEntry struct {
+	// prefix is the nibble path to the node in the trie
+	prefix []byte
+	// node is the stacked node
+	node codec.EncodedNode
+	// encodedNode is the encoded node data
+	encodedNode []byte
+	// nodeHash of the node or nil if the node is inlined
+	nodeHash *[]byte
+	// omitValue is a flag to know if the value should be omitted in the generated proof
+	omitValue bool
+	// childIndex is used for branch nodes
+	childIndex int
+	// children contains the child references to use in constructing the proof nodes.
+	children [codec.ChildrenCapacity]triedb.ChildReference
+	// outputIndex is the index into the proof vector that the encoding of this entry should be placed at.
+	outputIndex *int
+}
+
+func newStackEntry(
+	prefix []byte,
+	nodeData []byte,
+	nodeHash *[]byte,
+	outputIndex *int) (*stackEntry, error) {
+	node, err := codec.Decode(bytes.NewReader(nodeData))
+	if err != nil {
+		return nil, err
+	}
+
+	return &stackEntry{
+		prefix:      prefix,
+		node:        node,
+		encodedNode: nodeData,
+		nodeHash:    nodeHash,
+		omitValue:   false,
+		childIndex:  0,
+		children:    [codec.ChildrenCapacity]triedb.ChildReference{},
+		outputIndex: outputIndex,
+	}, nil
+}
+
+func (e *stackEntry) encodeNode() ([]byte, error) {
+	switch n := e.node.(type) {
+	case codec.Empty:
+		return e.encodedNode, nil
+	case codec.Leaf:
+		if !e.omitValue {
+			return e.encodedNode, nil
+		}
+
+		encodingBuffer := bytes.NewBuffer(nil)
+		err := triedb.NewEncodedLeaf(e.node.GetPartialKey(), codec.NewInlineValue([]byte{}), encodingBuffer)
+		if err != nil {
+			return nil, err
+		}
+
+		return encodingBuffer.Bytes(), nil
+	case codec.Branch:
+		var value codec.EncodedValue
+		if !e.omitValue {
+			value = n.Value
+		}
+		e.completBranchChildren(n.Children, e.childIndex)
+		encodingBuffer := bytes.NewBuffer(nil)
+		err := triedb.NewEncodedBranch(e.node.GetPartialKey(), e.children, value, encodingBuffer)
+		if err != nil {
+			return nil, err
+		}
+		return encodingBuffer.Bytes(), nil
+	default:
+		panic("unreachable")
+	}
+}
+
+func (e *stackEntry) setChild(encodedChild []byte) {
+	var childRef triedb.ChildReference
+	switch n := e.node.(type) {
+	case codec.Empty, codec.Leaf:
+		panic("Empty and leaf nodes have no children, we cannot descend into")
+	case codec.Branch:
+		if e.childIndex >= codec.ChildrenCapacity {
+			panic("child index out of bounds")
+		}
+		if n.Children[e.childIndex] != nil {
+			childRef = e.replaceChildRef(encodedChild, n.Children[e.childIndex])
+		}
```

Comment:
Can we use `default` here?
```suggestion
	case codec.Branch:
		if e.childIndex >= codec.ChildrenCapacity {
			panic("child index out of bounds")
		}
		if n.Children[e.childIndex] != nil {
			childRef = e.replaceChildRef(encodedChild, n.Children[e.childIndex])
		}
	default:
		panic("Empty and leaf nodes have no children, we cannot descend into")
```

Reply:
Sure! Done!